File: common/devices/osxbluetooth.c
/*

  $Id$

  G N O K I I

  A Linux/Unix toolset and driver for the mobile phones.

  This file is part of gnokii.

  Gnokii is free software; you can redistribute it and/or modify
  it under the terms of the GNU General Public License as published by
  the Free Software Foundation; either version 2 of the License, or
  (at your option) any later version.

  Gnokii is distributed in the hope that it will be useful,
  but WITHOUT ANY WARRANTY; without even the implied warranty of
  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
  GNU General Public License for more details.

  You should have received a copy of the GNU General Public License
  along with gnokii; if not, write to the Free Software
  Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA

  Copyright (C) 1999-2000  Hugh Blemings & Pavel Janík ml.
  Copyright (C) 2003       Siegfried Schloissnig

*/

#include "config.h"

#ifdef HAVE_BLUETOOTH_MACOSX

#include "compat.h"
#include "devices/osxbluetooth.h"
#include "misc.h"
#include "gnokii.h"
#include "devices/unixbluetooth.h"

/* ----- bluetooth io thread ----- */

static void thread_rfcommDataListener(IOBluetoothRFCOMMChannelRef rfcommChannel,
                               void* data, UInt16 length, void* refCon)
{
	threadContext *pContext = (threadContext *)refCon;
	void *pBuffer = malloc(length);
	dataBlock *pDataBlock = (dataBlock *)malloc(sizeof(dataBlock));

	memcpy(pBuffer, data, length);

	pDataBlock->pData = pBuffer;
	pDataBlock->nSize = length;

	pthread_mutex_lock(&(pContext->mutexWait));
	CFArrayAppendValue(pContext->arrDataReceived, pDataBlock);
	pthread_mutex_unlock(&(pContext->mutexWait));
}

static void *thread_main(void *pArg)
{
	threadContext* pContext = (threadContext *)pArg;
	IOBluetoothDeviceRef device = IOBluetoothDeviceCreateWithAddress(&(pContext->deviceAddress));
	IOBluetoothRFCOMMChannelRef rfcommChannel;

	if (IOBluetoothDeviceOpenRFCOMMChannel(device, pContext->nChannel,
				      &rfcommChannel) != kIOReturnSuccess) {
		rfcommChannel = 0;
	} else {
		/* register an incoming data listener */
		if (IOBluetoothRFCOMMChannelRegisterIncomingDataListener(rfcommChannel,
			 thread_rfcommDataListener, pArg) != kIOReturnSuccess) {
		    rfcommChannel = 0;
		}
	}

	pContext->rfcommChannel = rfcommChannel;

	pthread_mutex_unlock(&(pContext->mutexWait));

	/* start the runloop */
	CFRunLoopRun();
}

/* ---- bluetooth io thread ---- */

int bluetooth_open(const char* addr, uint8_t channel, struct gn_statemachine* state)
{
	/* create the thread context and start the thread */
	CFStringRef strDevice;
	pthread_t threadID;
	threadContext *pContext = (threadContext *)malloc(sizeof(threadContext));

	strDevice = CFStringCreateWithCString(kCFAllocatorDefault, addr, kCFStringEncodingMacRomanLatin1);
	IOBluetoothNSStringToDeviceAddress(strDevice, &pContext->deviceAddress);
	CFRelease(strDevice);

	pContext->arrDataReceived = CFArrayCreateMutable(kCFAllocatorDefault, 0, NULL);
	pContext->rfcommChannel = 0;
	pContext->nChannel = channel;

	pthread_mutex_init(&(pContext->mutexWait), NULL);
	pthread_mutex_lock(&(pContext->mutexWait));	/* lock */

	pthread_create(&(pContext->threadID), NULL, thread_main, pContext);

	/* wait until main finishes its initialization */
	pthread_mutex_lock(&(pContext->mutexWait));
	/* unblock the mutex */
	pthread_mutex_unlock(&(pContext->mutexWait));

	if (pContext->rfcommChannel == 0)
		return -1;
	else
		/* return the thread context as the file descriptor */
		return (int)pContext;
}

int bluetooth_close(int fd, struct gn_statemachine *state)
{
	threadContext *pContext = (threadContext *)fd;
        IOBluetoothDeviceRef device;

	sleep(2);

	if (fd != -1 && pContext->rfcommChannel > 0) {
		/* de-register the callback */
		IOBluetoothRFCOMMChannelRegisterIncomingDataListener(pContext->rfcommChannel, NULL, NULL);

		/* close channel and device connection */
		IOBluetoothRFCOMMChannelCloseChannel(pContext->rfcommChannel);
		device = IOBluetoothRFCOMMChannelGetDevice(pContext->rfcommChannel);
		IOBluetoothDeviceCloseConnection(device);
		/* IOBluetoothObjectRelease(pContext->rfcommChannel); */
		IOBluetoothObjectRelease(device);
	}

	return 1;
}

int bluetooth_write(int fd, const __ptr_t bytes, int size, struct gn_statemachine *state)
{
	threadContext *pContext = (threadContext *)fd;

	if (IOBluetoothRFCOMMChannelWrite(pContext->rfcommChannel, bytes, size, TRUE) != kIOReturnSuccess)
		return -1;

	return size;
}

int bluetooth_read(int fd, __ptr_t bytes, int size, struct gn_statemachine *state)
{
	threadContext *pContext = (threadContext *)fd;
	int nOffset = 0;
	int nBytes = 0;
        dataBlock* pDataBlock;

	/* no data received so far */
	if (CFArrayGetCount(pContext->arrDataReceived) == 0)
		return 0;

	while (CFArrayGetCount(pContext->arrDataReceived) != 0) {
		pthread_mutex_lock(&(pContext->mutexWait));
		pDataBlock = (dataBlock*)CFArrayGetValueAtIndex(pContext->arrDataReceived, 0);
		pthread_mutex_unlock(&(pContext->mutexWait));

		if (pDataBlock->nSize == size) {
			/* copy data and remove block */
			memcpy(bytes + nOffset, pDataBlock->pData, size);

			pthread_mutex_lock(&(pContext->mutexWait));
			CFArrayRemoveValueAtIndex(pContext->arrDataReceived, 0);
			pthread_mutex_unlock(&(pContext->mutexWait));

			free(pDataBlock->pData);
			free(pDataBlock);

			return nBytes + size;
		} else if (pDataBlock->nSize > size) {
			/* copy data and update block contents */
			memcpy(bytes + nOffset, pDataBlock->pData, size);
			memmove(pDataBlock->pData, pDataBlock->pData + size, pDataBlock->nSize - size);
			pDataBlock->nSize -= size;
			return nBytes + size;
		} else { /* pDataBlock->nSize < size */
			/* copy data and remove block */
			memcpy(bytes + nOffset, pDataBlock->pData, pDataBlock->nSize);

			size -= pDataBlock->nSize;
			nOffset += pDataBlock->nSize;
			nBytes += pDataBlock->nSize;

			pthread_mutex_lock(&(pContext->mutexWait));
			CFArrayRemoveValueAtIndex(pContext->arrDataReceived, 0);
			pthread_mutex_unlock(&(pContext->mutexWait));

			free(pDataBlock->pData);
			free(pDataBlock);
		}
	}

	return nBytes;
}

int bluetooth_select(int fd, struct timeval *timeout, struct gn_statemachine *state)
{
	threadContext *pContext = (threadContext *)fd;
	int nRetVal = 0;

	usleep(timeout->tv_usec);

	pthread_mutex_lock(&(pContext->mutexWait));

	if (CFArrayGetCount(pContext->arrDataReceived) > 0)
		nRetVal = 1;

	pthread_mutex_unlock(&(pContext->mutexWait));

	return nRetVal;
}

#endif	/* HAVE_BLUETOOTH_MACOSX */

File: common/devices/osxbluetooth.m
/*

  G N O K I I

  A Linux/Unix toolset and driver for the mobile phones.

  This file is part of gnokii.

  Copyright (C) 2015       Fabrizio Gennari

*/

#include "config.h"

#ifdef HAVE_BLUETOOTH_MACOSX

#include <IOBluetooth/objc/IOBluetoothRFCOMMChannel.h>
#include <IOBluetooth/objc/IOBluetoothDevice.h>

#include "devices/bluetooth.h"

static NSMutableDictionary *queues;
static int next_fd = 1;

@interface GnokiiOSXBluetooth : NSObject <IOBluetoothRFCOMMChannelDelegate> {
@private
    NSMutableData *queue;
    IOBluetoothRFCOMMChannel *channel;
}
- (BOOL)connect:(const char*)addr chid:(BluetoothRFCOMMChannelID)channelID;
- (int)select:(CFAbsoluteTime)timeoutDate;
- (int)read:(__ptr_t)bytes size:(int)size;
- (BOOL)write:(const __ptr_t *)data length:(int)length;
@end

@implementation GnokiiOSXBluetooth
- (id)init {
    if ((self = [super init])) {
        queue = [[NSMutableData alloc] init];
    }
    return self;
}

- (BOOL)connect:(const char*)addr chid:(BluetoothRFCOMMChannelID)channelID {
    NSString *address = [[[NSString alloc] initWithCString:addr encoding:NSASCIIStringEncoding] autorelease];
    IOBluetoothDevice* btDevice = [[IOBluetoothDevice deviceWithAddressString:address] autorelease];
    return btDevice != nil &&
    [btDevice openRFCOMMChannelSync:&channel withChannelID:channelID delegate:self] == kIOReturnSuccess; // after connection it is established.. the delegates methoed are triggered.
}

-(BOOL)write:(const __ptr_t *)data length:(int)length {
    int i;\
    IOReturn ret = [channel writeSync:(void*)data length:(UInt16)length];
    return ret == kIOReturnSuccess;
}

- (void)dealloc {
    if (channel != nil) {
        [channel closeChannel];
        [[channel getDevice] closeConnection];
        [channel release];
    }
    [queue release];
    [super dealloc];
}

- (int)waitForData:(NSTimeInterval)timeout {
    int ret = -1;
    if ([queue length] <= 0 && [channel isOpen]) {
        NSTimer *timer = nil;
        if (timeout > 0){
            timer =     [NSTimer scheduledTimerWithTimeInterval:timeout
             target:self
             selector:@selector(timerElapsed:)
             userInfo:nil
             repeats:NO];
        }
        CFRunLoopRun();
        if (timer != nil){
            if ([timer isValid])
                [timer invalidate];
            else
                ret = 0;
        }
    }
    if ([queue length] > 0)
        return 1;
    return ret;
}

- (int)select:(NSTimeInterval)timeout {
    return [self waitForData:timeout];
}

-(int)read:(__ptr_t)bytes size:(int)size {
    if (![self waitForData:0])
        return -1;

    if (size > [queue length])
        size = [queue length];
    memcpy (bytes, [queue mutableBytes], size);
    NSRange range = {0, size};
    [queue replaceBytesInRange:range withBytes:nil length:0];
    return size;
}

- (void)rfcommChannelData:(IOBluetoothRFCOMMChannel*)rfcommChannel data:(void *)dataPointer length:(size_t)dataLength {
    int i;
    [queue appendBytes:dataPointer length:dataLength];
    if ([queue length] > 0)
        CFRunLoopStop(CFRunLoopGetCurrent());
}

- (void)rfcommChannelClosed:(IOBluetoothRFCOMMChannel*)rfcommChannel {
    if (channel == rfcommChannel)
        CFRunLoopStop(CFRunLoopGetCurrent());
}
- (void) timerElapsed:(NSTimer*) timer {
    CFRunLoopStop(CFRunLoopGetCurrent());
}
@end

int bluetooth_open(const char* addr, uint8_t channel_num, struct gn_statemachine* state)
{
    if (queues == nil)
        queues = [NSMutableDictionary dictionaryWithCapacity:1];
    GnokiiOSXBluetooth *q = [[GnokiiOSXBluetooth alloc] init];
    if (![q connect:addr chid:(BluetoothRFCOMMChannelID)channel_num]) { // after connection it is established.. the delegates methoed are triggered.
        [q release];
        return -1;
    }
    int ret = next_fd++;
    [queues setObject:q forKey:@(ret)];
    return ret;
}

int bluetooth_write(int fd, const __ptr_t bytes, int size, struct gn_statemachine *state)
{
    if (queues == nil)
        return -1;
    GnokiiOSXBluetooth *q = [queues objectForKey:@(fd)];
    sleep(2);
    if (q == nil
        || ![q write:bytes length:size])
        return -1;

    return size;
}

int bluetooth_read(int fd, __ptr_t bytes, int size, struct gn_statemachine *state)
{
    if (queues == nil)
        return -1;
    GnokiiOSXBluetooth *q = [queues objectForKey:@(fd)];
    if (q == nil)
        return -1;
    return [q read:bytes size:size];
}

int bluetooth_select(int fd, struct timeval *timeout, struct gn_statemachine *state)
{
    if (queues == nil)
        return -1;
    GnokiiOSXBluetooth *q = [queues objectForKey:@(fd)];
    if (q == nil)
        return -1;

    return [q select:(NSTimeInterval)(timeout->tv_sec + timeout->tv_usec / 1000000.0)];
}

int bluetooth_close(int fd, struct gn_statemachine *state)
{
    if (queues == nil)
        return -1;
    GnokiiOSXBluetooth *q = [queues objectForKey:@(fd)];
    if (q == nil)
        return -1;
    [q release];
    [queues removeObjectForKey:@(fd)];
    return 1;
}

#endif
